File: bindings/dotnet/src/Aster/Interceptors/InterceptorChain.cs
namespace Aster.Interceptors;

/// <summary>
/// Applies a list of interceptors in order (request/response forward, error reverse).
/// </summary>
public static class InterceptorChain
{
    /// <summary>
    /// Run each interceptor's OnRequest in forward order.
    /// </summary>
    public static object ApplyRequest(IReadOnlyList<IInterceptor> interceptors, CallContext ctx, object request)
    {
        var current = request;
        foreach (var interceptor in interceptors)
        {
            current = interceptor.OnRequest(ctx, current);
        }
        return current;
    }

    /// <summary>
    /// Run each interceptor's OnResponse in forward order.
    /// </summary>
    public static object ApplyResponse(IReadOnlyList<IInterceptor> interceptors, CallContext ctx, object response)
    {
        var current = response;
        foreach (var interceptor in interceptors)
        {
            current = interceptor.OnResponse(ctx, current);
        }
        return current;
    }

    /// <summary>
    /// Run each interceptor's OnError in reverse order. Returns null if any interceptor suppresses the error.
    /// </summary>
    public static RpcError? ApplyError(IReadOnlyList<IInterceptor> interceptors, CallContext ctx, RpcError error)
    {
        RpcError? current = error;
        for (int i = interceptors.Count - 1; i >= 0; i--)
        {
            if (current is null)
                return null;
            current = interceptors[i].OnError(ctx, current);
        }
        return current;
    }
}

File: bindings/java/src/main/java/com/aster/interceptors/InterceptorChain.java
package com.aster.interceptors;

import java.util.List;

/**
 * Static helpers that drive a list of {@link Interceptor}s over a request/response/error lifecycle.
 *
 * <p>Request and response interceptors run in forward order. Error interceptors run in reverse
 * order (innermost first), mirroring the Python implementation.
 */
public final class InterceptorChain {

  private InterceptorChain() {}

  /** Applies {@code onRequest} in forward order, returning the final request object. */
  public static Object applyRequest(
      List<Interceptor> interceptors, CallContext ctx, Object request) {
    Object current = request;
    for (Interceptor interceptor : interceptors) {
      current = interceptor.onRequest(ctx, current);
    }
    return current;
  }

  /** Applies {@code onResponse} in forward order, returning the final response object. */
  public static Object applyResponse(
      List<Interceptor> interceptors, CallContext ctx, Object response) {
    Object current = response;
    for (Interceptor interceptor : interceptors) {
      current = interceptor.onResponse(ctx, current);
    }
    return current;
  }

  /**
   * Applies {@code onError} in reverse order, returning the final error or {@code null} if
   * suppressed.
   */
  public static RpcError applyError(
      List<Interceptor> interceptors, CallContext ctx, RpcError error) {
    RpcError current = error;
    for (int i = interceptors.size() - 1; i >= 0; i--) {
      if (current == null) {
        return null;
      }
      current = interceptors.get(i).onError(ctx, current);
    }
    return current;
  }
}
